src/python/pants/source/source_root.py

```diff
@@ -202,31 +229,55 @@ class OptionalSourceRoot:
 
 
 @rule
-def get_source_root(
+async def get_source_root(
     source_root_request: SourceRootRequest, source_root_config: SourceRootConfig
 ) -> OptionalSourceRoot:
     """Rule to request a SourceRoot that may not exist."""
-    matched_path = source_root_config.get_pattern_matcher().find_root(source_root_request.path)
-    if matched_path:
-        return OptionalSourceRoot(SourceRoot(path=str(matched_path)))
-    else:
-        return OptionalSourceRoot(None)
+    pattern_matcher = source_root_config.get_pattern_matcher()
+    path = source_root_request.path
+
+    # Check if the requested path itself is a source root.
+
+    # A) Does it match a pattern?
+    if pattern_matcher.matches_root_patterns(path):
+        return OptionalSourceRoot(SourceRoot(str(path)))
+
+    # B) Does it contain a marker file?
+    marker_filenames = source_root_config.options.marker_filenames
+    if marker_filenames:
+        for marker_filename in marker_filenames:
+            if (
+                os.path.basename(marker_filename) != marker_filename
+                or "*" in marker_filename
+                or "!" in marker_filename
+            ):
+                raise InvalidMarkerFileError(
+                    f"Marker filename must be a base name: {marker_filename}"
+                )
+        # TODO: An intrinsic to check file existence at a fixed path?
+        snapshot = await Get[Snapshot](PathGlobs([str(path / mf) for mf in marker_filenames]))
```

Comment:
you can use a generator expression here instead of a list comprehension since PathGlobs will sort it and convert it into a tuple anyway.